service/storage_proxy.cc

```diff
@@ -1726,11 +1726,17 @@ class storage_proxy::cancellable_write_handlers_list : public bi::list<abstract_
         // handler is being removed from the b::list, so if any live iterator points at it,
         // move it to the next object (this requires that the list is traversed in the forward
         // direction).
+        bool drop_end = false;
         for (auto& itp : _live_iterators) {
             if (&**itp == handler) {
                 ++*itp;
+                drop_end |= (*itp == end());
             }
         }
+        if (drop_end) {
+            const auto [first, last] = std::ranges::remove_if(_live_iterators, [this] (iterator* pit) { return *pit == end(); });
+            _live_iterators.erase(first, last);
+        }
```

Comment:
Alternative solution: document that iterator_guard/register_live_iterator must not be called on an end iterator, and change the order in cancel_write_handlers:


```diff
-        ++it;
         if (need_preempt()) {
             cancellable_write_handlers_list::iterator_guard ig{*_cancellable_write_handlers_list, it};
             seastar::thread::yield();
         }
+        ++it;
```

Reply:
Ah, it's not sufficient because an iterator could become the end iterator while the loop is preempted.